guppylang/checker/func_checker.py

```diff
@@ -148,8 +148,13 @@ def check_signature(func_def: ast.FunctionDef, globals: Globals) -> FunctionType
     for inp in func_def.args.args:
         if inp.annotation is None:
             raise GuppyError("Argument type must be annotated", inp)
-        ty = type_from_ast(inp.annotation, globals, param_var_mapping)
-        inputs.append(FuncInput(ty, InputFlags.NoFlags))
+        ty, flags = type_with_flags_from_ast(inp.annotation, globals, param_var_mapping)
```

Comment:
This does seem a bit like, we have two places that define the set(s) of flags legal in the signature of a function - `_CallableTypeDef` also makes similar checks. If you parsed `type_with_flags_from_ast` for the return type too, you could call some single checking function (maybe it takes a `list[FlaggedArg]` and a single `FlaggedArg` and then `_CallableTypeDef` first separates out the last element of the list) that issues appropriate errors and returns some appropriate e.g. `(list[TypeArg],TypeArg)`